### PR TITLE
Bump CUDA from 11.8 to 12.0

### LINKED ISSUE #489: Flax needs to be upgraded in the tensorflow/jax image
**Describe the bug**
Current version of flax (0.6.1) on the image does not work properly with the jax version (0.4.13). 
T

**To Reproduce**
Issue can be reproduced by doing `from flax.training import checkpoints`, which will give the error `ModuleNotFoundError: No module named 'jax.experimental.global_device_array'`. 
This has been discussed in https://github.com/google/flax/issues/3087. 

**Expected behavior**
Flax should be importable.

**Infrastructure (Where you are running this image):**
- LEAP hub (@jbusecke)

**Solution**
At the moment I solve this doing
`pip install flax==0.6.10`

### PR DESCRIPTION
**Pull request by weiji14**
CUDA 12.0 migration across conda-forge is practically complete (see https://conda-forge.org/status/#cuda120), so we can start updating to a newer version of CUDA!

Note:
- CUDA 12.x requires CUDA driver >=525.60.13, see https://docs.nvidia.com/deploy/cuda-compatibility/index.html#cuda-intro and https://docs.nvidia.com/cuda/archive/12.0.0/cuda-toolkit-release-notes/index.html#cuda-toolkit-major-component-versions. Previous CUDA 11.x required CUDA driver >= 450.80.02.

Changes in this PR:
- [x] Update Pytorch, Torchvision and Tensorflow to use CUDA 12.0 builds
- [x] Update minimum pin on Tensorflow from >=2.14.0 to >=2.15.0
- [x] Update minimum pin on Pytorch from 2.0.0 to 2.1.0, and torchvision from 0.15.1 to 0.16.1
- [x] Pin minimum version of flax to 0.8.0, helps to resolve #489

References:
- https://developer.nvidia.com/blog/cuda-toolkit-12-0-released-for-general-availability
- https://github.com/conda-forge/conda-forge.github.io/issues/1963
- https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4400

Supersedes #505, Fixes #489

### CODE REVIEW

**Comment by weiji14** [comment]
Hmm, conda-lock is not handling the `__cuda` constraint somehow, even though I set the `CONDA_OVERRIDE_CUDA` environment variable already at 4d26a191867cf378f73588c34e2a1bce0900ad06:

```
The following package could not be installed
└─ tensorflow >=2.15.0 *cuda120* is not installable because it requires
   ├─ __cuda, which can be installed;
   └─ tensorflow-estimator [2.15.0 cuda120py310h549c77d_2|2.15.0 cuda120py310h549c77d_3|...|2.15.0 cuda120py39ha585809_3], which requires
      └─ cuda-version >=12.0,<13 , which requires
         └─ __cuda >=12 , which conflicts with any installable versions previously reported.
```
Need to see what's going on.

**Comment by weiji14** [comment]
For anyone interested, this PR is still stuck because conda-lock doesn't want to solve the `ml-notebook`/`pytorch-notebook` conda environment, keep getting the same error mentioned at https://github.com/pangeo-data/pangeo-docker-images/pull/514#issuecomment-1947710186.

**Comment by weiji14** [comment]
Update: Finally, after setting a `cuda-version=12.0` pin (instead of setting it in the build version), the `ml-notebook` locking worked at https://github.com/pangeo-data/pangeo-docker-images/actions/runs/9132472927/job/25113889544#step:4:20 🎉 . The locking on `pytorch-notebook` timed out after 20min though, so increasing that to 30min and trying again 🤞

**Comment by weiji14** [comment]
Ah ok, can't just change the timeout in this PR, might need to have it on the `master` branch first.

**Comment by weiji14** [comment]
Oh wow, now the locking takes just 2min 30s :sweat_smile: Anyways, this is finally ready for review!

File: pytorch-notebook/conda-lock.yml [hunk 1]
```diff
@@ -9209,35 +9118,29 @@ package:
...
-    nccl: '>=2.21.5.1,<3.0a0'
     networkx: ''
     numpy: '>=1.23.5,<2.0a0'
     python: '>=3.11,<3.12.0a0'
     python_abi: 3.11.*
     sleef: '>=3.5.1,<4.0a0'
     sympy: ''
     typing_extensions: ''
-  url: https://conda.anaconda.org/conda-forge/linux-64/pytorch-2.3.0-cuda118_py311h6c9cb27_300.conda
+  url: https://conda.anaconda.org/conda-forge/linux-64/pytorch-2.3.0-cpu_mkl_py311h9835ca6_100.conda
```
Comment (weiji14) [inline pytorch-notebook/conda-lock.yml:9140]: Hmm, the cpu version is getting pulled in, might need to actually do the `cuda*` pin on the build number still...

**Comment by weiji14** [comment]
Found a clue after creating a [`virtual-packages.yml`](https://github.com/conda/conda-lock/tree/v2.5.7?tab=readme-ov-file#--virtual-package-spec) file with the following contents:

```
subdirs:
  linux-64:
    packages:
      __cuda: "12.0"
```

Running `conda-lock lock -f environment.yml -f ../pangeo-notebook/environment.yml -f ../base-notebook/environment.yml -p linux-64` gave this error:

```
INFO:conda_lock.conda_lock:Using virtual packages from virtual-packages.yml
Locking dependencies for ['linux-64']...
INFO:conda_lock.conda_solver:linux-64 using specs ['cuda-version 12.0.*', 'jupyterlab-nvdashboard', 'gpytorch', 'pytorch >=2.3.0 cuda120*', 'torchvision >=0.18.0 cuda120*', 'torchgeo', 'adlfs', 'argopy', 'awscli', 'black', 'boto3', 'bottleneck', 'cartopy', 'cdsapi', 'cfgrib', 'cf_xarray', 'ciso', 'cmocean', 'dask-ml', 'datashader', 'descartes', 'earthaccess', 'eofs', 'erddapy', 'esmpy', 'fastjmd95', 'flox', 'fsspec', 'gcm_filters', 'gcsfs', 'gh', 'gh-scoped-creds', 'geocube', 'geopandas', 'geopy', 'geoviews-core', 'git-lfs', 'gsw', 'h5netcdf', 'h5py', 'holoviews', 'hvplot', 'intake', 'intake-esm', 'intake-geopandas', 'intake-stac', 'intake-xarray', 'ipdb', 'ipykernel', 'ipyleaflet', 'ipytree', 'ipywidgets', 'jupyterlab_code_formatter', 'jupyterlab-git', 'jupyterlab-lsp', 'jupyterlab-myst', 'jupyter-panel-proxy', 'jupyter-resource-usage', 'kerchunk', 'line_profiler', 'lxml', 'lz4', 'matplotlib-base', 'memory_profiler', 'metpy', 'nb_conda_kernels', 'nbstripout', 'nc-time-axis', 'netcdf4', 'numbagg', 'numcodecs', 'numpy', 'numpy_groupies', 'odc-stac', 'pandas', 'panel', 'parcels', 'param', 'pop-tools', 'pyarrow', 'pycamhd', 'pydap', 'pystac', 'pystac-client', 'python-blosc', 'python-gist', 'python-graphviz', 'python-lsp-ruff', 'python-xxhash', 'rasterio', 'rechunker', 'rio-cogeo', 'rioxarray', 'ruff', 's3fs', 'satpy', 'scikit-image', 'scikit-learn', 'scipy', 'seaborn', 'sparse', 'snakeviz', 'stackstac', 'tiledb-py', 'timezonefinder', 'watermark', 'xarray', 'xarrayutils', 'xarray-datatree', 'xarray_leaflet', 'xarray-spatial', 'xbatcher', 'xcape', 'xclim', 'xesmf', 'xgboost', 'xgcm', 'xhistogram', 'xmip', 'xmitgcm', 'xpublish', 'xrft', 'xskillscore', 'xxhash', 'zarr', 'python 3.11.*', 'pangeo-notebook 2024.05.20.*', 'pip']
Failed to parse json, Expecting value: line 1 column 1 (char 0)
Could not lock the environment for platform linux-64
Could not solve for environment specs
The following packages are incompatible
├─ cuda-version 12.0**  is requested and can be installed;
└─ torchvision >=0.18.0 cuda120* is not installable because it requires
   └─ libcusparse >=12.3.1.170,<13.0a0 , which requires
      └─ cuda-version >=12.4,<12.5.0a0 , which conflicts with any installable versions previously reported.
{
    "success": false
}
```
Need to play around with relaxing the `cuda-version` pinning a bit.

File: pytorch-notebook/conda-lock.yml [hunk 2]
```diff
@@ -9234,10 +9394,10 @@ package:
     sleef: '>=3.5.1,<4.0a0'
     sympy: ''
     typing_extensions: ''
-  url: https://conda.anaconda.org/conda-forge/linux-64/pytorch-2.3.0-cuda118_py311h6c9cb27_300.conda
+  url: https://conda.anaconda.org/conda-forge/linux-64/pytorch-2.3.0-cuda120_py311h2667f23_300.conda
```
Comment (weiji14) [inline pytorch-notebook/conda-lock.yml:9397]: Yes, finally getting the cuda120 build version of pytorch!!!

File: pytorch-notebook/conda-lock.yml (hunk 1 above)
Comment (weiji14) [reply pytorch-notebook/conda-lock.yml]: Update: `pytorch-2.3.0-cuda120*` is pulled in now, see https://github.com/pangeo-data/pangeo-docker-images/pull/514#discussion_r1607658979

**Comment by scottyhq** [comment]
Thanks for sticking with this @weiji14 ! Seems like adding `vitural-packages.yml` was key, although maybe it's just a matter of new releases with more compatible dependencies. In any case, this also reduces the ML image size by ~700MB ! 

tensorflow: 11.3 -> 10.7GB
pytorch: 13.7 -> 13 GB

I'm going to go ahead and merge.

**Review by scottyhq — APPROVED** [review]

**Comment by weiji14** [comment]
Yeah, I want to say that the `virtual-packages.yml` was the thing that got it to work, but still confused as to why `conda-lock`'s `--with-cuda` flag didn't work. It's hard to create a minimal reproducible example to report upstream, and I'm not sure which packages are nudging the solver to an 'incorrect' solution with cpu packages, so I'll just leave it at that for now.

> In any case, this also reduces the ML image size by ~700MB !

Yes, this is because conda-forge has split the `cudatoolkit` package into different sub-components for CUDA 12 (see https://github.com/conda-forge/conda-forge.github.io/issues/1963) like `libcublas-dev`, `libcusparse-dev`, etc, so should be much lightweight now.

**Pull request merged**